### PR TITLE
Pokemon jump

### PR DESCRIPTION
**Pull request by PikalaxALT**
Ports Pokemon Jump and splits into 6 files according to section rearrangement between the two releases.

### CODE REVIEW

**Comment by jiangzhengwenjz** [comment]
see pret/pokeemerald#982

File: include/pokemon_jump.h [hunk 1]
```diff
@@ -2,9 +2,189 @@
...
+    u8 unkE;
+    u8 unkF;
+    u16 filler10;
+    u16 unk12;
+    u16 unk14;
+    u32 unk18;
+    u16 unk1C[5];
+    u8 txtBuff[2][0x40];
+    u8 strBuff[0x100];
+    u16 tilemapBuffer[(0x81a8 - 0x1a6) / 2]; // 0x1A6
```
Comment (jiangzhengwenjz) [inline include/pokemon_jump.h]: `u16 tilemapBuffer[0x4000];`

File: src/pokemon_jump_3.c [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+void sub_8149A6C(struct PokemonJump1_MonInfo *arg0)
+{
+    struct MonInfoPacket packet;
+    packet.id = 1,
+    packet.species = arg0->species,
+    packet.otId = arg0->otId,
+    packet.personality = arg0->personality,
+        RfuPrepareSend0x2f00(&packet);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_3.c]: What's the purpose of the extra indent? 
And I don't see the advantage of using comma here

File: src/pokemon_jump_6.c [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        sub_814B5C4(data[1]);
+        CopyWindowToVram(data[1], 3);
+        data[0]++;
+        break;
+    case 1:
+        if (!IsDma3ManagerBusyWithBgCopy())
+            data[0]++;
+        break;
+    case 2:
+        if (gMain.newKeys & (A_BUTTON | B_BUTTON))
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_6.c]: `JOY_NEW`

File: src/pokemon_jump_2.c [hunk 4]
```diff
@@ -0,0 +1,1970 @@
...
+        gUnknown_203F3D4->unk82E4[i].unk4 = 0x7FFFFFFF;
+        gUnknown_203F3D4->unk82E4[i].unk14 = 0;
+        gUnknown_203F3D4->unk8B[i] = 9;
+    }
+}
+
+static s16 GetPokemonJumpSpeciesIdx(u16 species)
+{
+    u32 i;
+    for (i = 0; i < ARRAY_COUNT(gPkmnJumpSpecies); i++)
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `NELEMS`

File: src/pokemon_jump_2.c [hunk 5]
```diff
@@ -0,0 +1,1970 @@
...
+                    gUnknown_203F3D4->unk1C = 82;
+            }
+        }
+    }
+    else
+    {
+        if (!(gUnknown_203F3D4->unk50 & 8))
+        {
+            gUnknown_203F3D4->unk28 = gUnknown_846B694[gUnknown_203F3D4->unk50] + (gUnknown_203F3D4->unk51 * 7);
+            gUnknown_203F3D4->unk4E = gUnknown_846B6A4[sub_8149194() % ARRAY_COUNT(gUnknown_846B6A4)] + 2;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `NELEMS`

File: src/pokemon_jump_2.c [hunk 6]
```diff
@@ -0,0 +1,1970 @@
...
+
+static void sub_8149900(u16 arg0, u16 *arg1, u16 *arg2)
+{
+    *arg2 = arg0 >> 12;
+    *arg1 = arg0 & 0xFFF;
+}
+
+static u16 sub_8149910(void)
+{
+    u16 index = Random() % ARRAY_COUNT(gUnknown_846B764);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `NELEMS`

File: src/pokemon_jump_2.c [hunk 7]
```diff
@@ -0,0 +1,1970 @@
...
+};
+
+void StartPokemonJump(u16 partyIndex, MainCallback callback)
+{
+    u8 taskId;
+
+    if (gReceivedRemoteLinkPlayers)
+    {
+        gUnknown_203F3D4 = Alloc(sizeof(*gUnknown_203F3D4));
+        if (gUnknown_203F3D4)
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `!= NULL`

File: src/pokemon_jump_2.c [hunk 8]
```diff
@@ -0,0 +1,1970 @@
...
+        {
+            ResetTasks();
+            taskId = CreateTask(sub_8147DA0, 1);
+            gUnknown_203F3D4->unk8 = 0;
+            gUnknown_203F3D4->returnCallback = callback;
+            gUnknown_203F3D4->unk4 = taskId;
+            gUnknown_203F3D4->unk6 = GetMultiplayerId();
+            sub_8147D2C(&gUnknown_203F3D4->unk82A8[gUnknown_203F3D4->unk6], &gPlayerParty[partyIndex]);
+            sub_8147B60(gUnknown_203F3D4);
+            SetWordTaskArg(taskId, 2, (u32)gUnknown_203F3D4);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `(uintptr_t)`

File: src/pokemon_jump_2.c [hunk 9]
```diff
@@ -0,0 +1,1970 @@
...
+            }
+        }
+        break;
+    }
+}
+
+static void sub_8148F5C(TaskFunc func, u8 taskPriority)
+{
+    u8 taskId = CreateTask(func, taskPriority);
+    SetWordTaskArg(taskId, 14, (u32)gUnknown_203F3D4);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_2.c]: `uintptr_t`

File: src/pokemon_jump_5.c [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+
+static const u16 gPkmnJumpPal1[] = INCBIN_U16("graphics/link_games/pkmnjump_pal1.gbapal");
+static const u16 gPkmnJumpPal2[] = INCBIN_U16("graphics/link_games/pkmnjump_pal2.gbapal");
+
+static const u32 gPkmnJumpRopeGfx1[] = INCBIN_U32("graphics/link_games/pkmnjump_rope1.4bpp.lz");
+static const u32 gPkmnJumpRopeGfx2[] = INCBIN_U32("graphics/link_games/pkmnjump_rope2.4bpp.lz");
+static const u32 gPkmnJumpRopeGfx3[] = INCBIN_U32("graphics/link_games/pkmnjump_rope3.4bpp.lz");
+static const u32 gPkmnJumpRopeGfx4[] = INCBIN_U32("graphics/link_games/pkmnjump_rope4.4bpp.lz");
+
+static const u32 gPkmnJumpStarGfx[] = INCBIN_U32("graphics/link_games/pkmnjump_star.4bpp.lz");
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: static symbol should start with `s`

File: src/pokemon_jump_5.c [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+    }
+
+    arg0->unk81A8[multiplayerId] = NULL;
+}
+
+void sub_814AF0C(struct PokemonJump2 *arg0, int multiplayerId)
+{
+    sub_814ADB4(arg0->unk81BC[multiplayerId]);
+    arg0->unk81BC[multiplayerId]->data[7] = arg0->unk81A8[multiplayerId] - gSprites;
+    arg0->unk81BC[multiplayerId]->invisible = 0;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: `FALSE`

File: src/pokemon_jump_5.c [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+}
+
+static void sub_814AF74(struct Sprite *sprite)
+{
+    switch (sprite->data[0])
+    {
+    case 0:
+        if (sprite->animEnded)
+        {
+            sprite->invisible = 1;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: `TRUE`

File: src/pokemon_jump_5.c [hunk 13]
```diff
@@ -0,0 +1,637 @@
...
+        if (sprite->pos1.y <= 72)
+        {
+            sprite->pos1.y = 72;
+            sprite->data[0]++;
+        }
+        break;
+    case 2:
+        if (++sprite->data[1] >= 48)
+        {
+            sprite->invisible = 1;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: `TRUE`

File: src/pokemon_jump_5.c [hunk 14]
```diff
@@ -0,0 +1,637 @@
...
+
+void sub_814B0A8(struct PokemonJump2 *arg0)
+{
+    int i;
+    u16 numPlayers = sub_81499A4();
+    for (i = 0; i < numPlayers; i++)
+    {
+        if (arg0->unk81A8[i]->callback == sub_814B100)
+        {
+            arg0->unk81A8[i]->invisible = 0;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: `FALSE`

File: src/pokemon_jump_5.c [hunk 15]
```diff
@@ -0,0 +1,637 @@
...
+        break;
+    }
+}
+
+void sub_814B240(struct PokemonJump2 *arg0, s16 x, s16 y, u8 multiplayerId)
+{
+    u8 spriteId = CreateSprite(&gUnknown_846E2B4, x, y, 1);
+    if (spriteId != MAX_SPRITES)
+    {
+        gSprites[spriteId].invisible = 1;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_5.c]: `TRUE`

File: src/pokemon_jump_4.c [hunk 16]
```diff
@@ -0,0 +1,922 @@
...
+
+void sub_8149CEC(struct PokemonJump2 *arg0)
+{
+    u8 taskId;
+
+    gUnknown_203F3D8 = arg0;
+    sub_8149D34(gUnknown_203F3D8);
+    taskId = CreateTask(sub_8149DA4, 3);
+    gUnknown_203F3D8->unk6 = taskId;
+    SetWordTaskArg(gUnknown_203F3D8->unk6, 2, (u32) gUnknown_203F3D8);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: `(uintptr_t)`, no space following cast

File: src/pokemon_jump_4.c [hunk 17]
```diff
@@ -0,0 +1,922 @@
...
+}
+
+bool32 sub_8149D68(void)
+{
+    return (gUnknown_203F3D8->unk0 != 1);
+}
+
+static void sub_8149D80(void (*func)(void))
+{
+    SetWordTaskArg(gUnknown_203F3D8->unk6, 0, (u32) func);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: `(uintptr_t)`, no space following cast

File: src/pokemon_jump_4.c [hunk 18]
```diff
@@ -0,0 +1,922 @@
...
+static const u32 gPkmnJumpBgGfx[] = INCBIN_U32("graphics/link_games/pkmnjump_bg.4bpp.lz");
+static const u32 gPkmnJumpBgTilemap[] = INCBIN_U32("graphics/link_games/pkmnjump_bg.bin.lz");
+
+static const u16 gPkmnJumpVenusaurPal[] = INCBIN_U16("graphics/link_games/pkmnjump_venusaur.gbapal");
+static const u32 gPkmnJumpVenusaurGfx[] = INCBIN_U32("graphics/link_games/pkmnjump_venusaur.4bpp.lz");
+static const u32 gPkmnJumpVenusaurTilemap[] = INCBIN_U32("graphics/link_games/pkmnjump_venusaur.bin.lz");
+
+static const u16 gPkmnJumpResultsPal[] = INCBIN_U16("graphics/link_games/pkmnjump_results.gbapal");
+static const u32 gPkmnJumpResultsGfx[] = INCBIN_U32("graphics/link_games/pkmnjump_results.4bpp.lz");
+static const u32 gPkmnJumpResultsTilemap[] = INCBIN_U32("graphics/link_games/pkmnjump_results.bin.lz");
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: static symbol should start with `s`

File: src/pokemon_jump_4.c [hunk 19]
```diff
@@ -0,0 +1,922 @@
...
+    SetWordTaskArg(gUnknown_203F3D8->unk6, 0, (u32) func);
+    gUnknown_203F3D8->unk4 = 0;
+    gUnknown_203F3D8->unk0 = 0;
+}
+
+static void sub_8149DA4(u8 taskId)
+{
+    if (!gUnknown_203F3D8->unk0)
+    {
+        void (*func)(void) = (void *)(GetWordTaskArg(taskId, 0));
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: Do we really need this temp variable?

File: src/pokemon_jump_4.c [hunk 20]
```diff
@@ -0,0 +1,922 @@
...
+        if (gUnknown_203F3D8->unkC != 7)
+            break;
+    case 1:
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+const u8 gUnknown_846D94C[] = _("IES");
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: Is it static

File: src/pokemon_jump_4.c [hunk 21]
```diff
@@ -0,0 +1,922 @@
...
+    window.bg = 0;
+    window.tilemapLeft = left;
+    window.tilemapTop = top;
+    window.width = width;
+    window.height = height;
+    window.paletteNum = 0xF;
+    window.baseBlock = 0x43;
+
+    windowId = AddWindow(&window);
+    FillWindowPixelBuffer(windowId, 0x11);
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: `PIXEL_FILL`

File: src/pokemon_jump_4.c [hunk 22]
```diff
@@ -0,0 +1,922 @@
...
+}
+
+static void sub_814A9C8(void)
+{
+    struct DigitObjUtilTemplate unkStruct;
+    struct DigitObjUtilTemplate *ptr = &unkStruct; // This temp variable is needed to match, don't ask me why.
+
+    ptr->shape = SPRITE_SHAPE(8x8);
+    ptr->size = SPRITE_SIZE(8x8);
+    ptr->strConvMode = 0;
```
Comment (jiangzhengwenjz) [inline src/pokemon_jump_4.c]: Does `STR_CONV_MODE_LEFT_ALIGN` apply here?

File: src/pokemon_jump_3.c (hunk 2 above)
Comment (PikalaxALT) [reply src/pokemon_jump_3.c]: Ugh... When pressing cmd+tab on my mac, sometimes the "cmd" doesn't take before the "tab", and this happens.

File: src/pokemon_jump_4.c (hunk 22 above)
Comment (PikalaxALT) [reply src/pokemon_jump_4.c]: It doesn't look to be the same as with ConvertIntToDecimalStringN

**Pull request merged**